tools/kraken.py

```diff
@@ -55,51 +55,67 @@ def build(self, db, options=None, option_string=None):
         self.execute('kraken-build', db, db, options=options,
                      option_string=option_string)
 
-    def pipeline(self, inBam, db, outReport=None, outReads=None, filterThreshold=None, numThreads=None):
-
-        with util.file.fifo(2) as (fastq1_pipe, fastq2_pipe):
-            # do not convert this to samtools bam2fq unless we can figure out how to replicate
-            # the clipping functionality of Picard SamToFastq
-            picard = tools.picard.SamToFastqTool()
-            picard_opts = {
-                'CLIPPING_ATTRIBUTE': tools.picard.SamToFastqTool.illumina_clipping_attribute,
-                'CLIPPING_ACTION': 'X'
-            }
-            picard.execute(inBam, fastq1_pipe, fastq2_pipe,
-                           picardOptions=tools.picard.PicardTools.dict_to_picard_opts(picard_opts),
-                           JVMmemory=picard.jvmMemDefault,
-                           background=True)
-
-            opts = {
-                '--threads': util.misc.sanitize_thread_count(numThreads),
-            }
+    def pipeline(self, db, inBams, outReports=None, outReads=None, filterThreshold=None, numThreads=None):
+        assert outReads is not None or outReports is not None
+
+        n_bams = len(inBams)
+        # 2n for paired fastq, 1n for kraken output
+        n_pipes = n_bams * 3
+
+        threads = util.misc.sanitize_thread_count(numThreads)
+
```

Comment:
Would it make sense to `kraken --preload` before processing all the samples -- e.g., if one sequential reading is better than all the random reads? This is what I've been doing on AWS before calling `snakemake ... all_kraken`, as I think you had suggested.

Reply:
My plan is to read/load the kraken database from pipe if the config is an S3 location, which is equivalent to `--preload`. Otherwise, it's kind of heuristic because not preloading allows you to go faster (only partially load db) if you only hit a small subset of kmers, especially if the backing disk is SSD. But if the backing disk is HDD and you have large inputs (like with `st1` EBS instances, then `--preload` will be a huge win.

Reply:
Ah, yeah, I have the database on an `st1` type so it makes sense for me, but I see your point if it's on SSD.